cirq/api/google/v2/result.proto

```diff
@@ -0,0 +1,103 @@
+
+  // The results along with the parameters that generated these results.
+  // These represent the expanded parameters defined int he ParameterSweep
+  // which this SweepResult corresponds to.
+  repeated ParameterizedResult parameterized_results = 3;
+}
+
+// The parameters used to generate result along with the results for this
+// set of parameters.
+message ParameterizedResult {
+  // The parameter dict that was used when generating these results.
+  ParameterDict params = 1;
+
+  // The measurement results. This is a packed representation of all of the
+  // measurements for fixed set of parameters (specified by params above)
+  // across all of the repetitions for the set of parameters given the
+  // params field.
+  //
+  // The result of a measurement gate is a list of bits, one for each qubit
+  // that has been measured. Every Program has a fixed set of measurement gates.
+  // These gates are labeled by a key (this key is unique across the Program).
+  // The results of a single run of the program should consist of bits for
+  // each of the measurement gates. To order these bits we use the
+  // measurement_keys field of the SweepResult. This is an ordered list of
+  // measurement keys, each of which includes a list of qubits that the
+  // measurement acts on. We use the ordering of the measurement_keys and then
+  // the ordering of the qubits to define the order of the bit string for a
+  // single run of the measurement.
+  //
+  // If the keys are (k_i), for i \in {0, 1, ..., m-1}, and the measurement
+  // acts on qubits q_{i,0}, q_{i,1}, ..., q_{i,n_i-1} (so that ith key
+  // corresponds to n_i qubits), then the measurements bit results are
+  // defined as
+  //  r_0[0] r_0[1] ... r_0[n_i-1] r_1[0] r_1[1] ... r_1[n_1-1] ...
+  //   ... r_{m-1}[0] r_{m-1}[1] ... r_{m-1}[n_{m-1}-1]
+  // Here r_i are the measurement result for the ith key (order defined by
+  // measurement keys).  Since the ith key has n_i qubits, r_i is a length
+  // n_i bit string, and r_i[j] is the jth bit in this string (order
+  // following the list order of the qubits).
+  //
+  // The above describes a bit string for a single run of a program with
+  // fixed parameters. This program however may be repeated, the number
+  // of times the program was run for these parameters is defined by the
+  // num_repetitions field of the SweepResult. If R_l is the bit string
+  // for the lth repetition of the program (defined by the r_0[0]... bit
+  // string above), then the full results is the concatenation of these
+  // bit strings
+  //   R_0 R_1 ... R_{num_repetitions - 1}
+  //
+  // Finally this entire bit string is encoded into the bytes of this field
+  // using little endian notation. That is, the least significant bit of the
+  // bytes is the first bit of the bit string, the second-least significant
+  // bit of the bytes is the second bit of the bit string, etc.
+  bytes measurement_results = 2;
```

Comment:
We are talking about about 100 measurement bits, right?  This seems overly complex considering that.  Having a repeated QubitResult {measurement_key, bytes repetition_results} for each bit would only add a little overhead and make things more clear.

Reply:
100 measurement bits times 10000 repetitions.  If you do repeated QubitResult {key, bytes results} then you still have to compress along the byte results or you can lose factor of 8 in size (of course over wire this likely gets compressed).  

I think we should keep until we really understand the size of results we have.

Reply:
You could have the key once per qubit and then have a bytes for each repetition, so only have 100 keys, not 100*repetitions.

I could just see a lot of lost time here if you accidentally unpack the results of the qubits in the wrong order and get weird results for a long time before you realize why.  It would be an error that would be really really hard to debug.

Reply:
agreed.  though we have written the code in cirq and elsewhere to do this multiple times now which does this packing and unpacking. 

note that the reason I am a real pain about the size of these things is that I believe that because you are not acting as an in memory processor right beside the quantum computer, the most important way to design efficiently against the engine api is to pack as much into a single job.  This means that a primary hit to your speed will be related to the size of this (storage but also serialization deserialization).  Otherwise I would say simplicity should rule.

Reply:
I am not sure I agree with this approach.  If you are doing 100 bits with 10000 repetitions, then this will take at least 1ms * 10000 = 10 seconds, so I highly doubt adding a few bytes per measurement bit is going to matter to the total latency of 10 seconds.

Contrast that to the proposed approach.  Due to the (admittedly, highly efficient) packing of bits.  The documentation alone is over 50 lines, and I am not sure I 100% fully understand the details of it.  If you get any of it wrong, you introduce a bug that is literally impossible to detect (i.e. how do you detect that you have re-ordered the measurement of qubit 1 and 2 or if the results are noisy?)

Furthermore, according to strilanc@ ("there would be an argument order specific to the gate type") you have the potential to re-introduce this undetectable error every time you introduce a new gate type.

I highly recommend that we add a small amount of metadata (ie. one field per measurement gate or one field per measurement bit) as a sort of check-sum.  Otherwise, it's going to be absurdly difficult to make sure each component that touches the proto encodes the result consistently.

Reply:
I'm convinced :) I think the thing that needs to be compressed is "num repetitions" since size of parameterization or # qubits will be smaller.

Reply:
Woo woo!  I like the new way much better!